File: extractors/typescript/src/paths.ts
import { dirname, isAbsolute, relative, resolve, sep } from "node:path";

/**
 * Paths as the model writes them: `/`-separated on every OS, root-relative,
 * compared by code unit. The file table is sorted by these strings, and a
 * Windows `\` or a case-folding comparison would change every surrogate.
 */

export function slashes(path: string): string {
  return path.replaceAll("\\", "/");
}

/** The typed path with `/` separators and no trailing slash — the header's `root`. */
export function display(typed: string): string {
  let text = slashes(typed);
  while (text.length > 1 && text.endsWith("/")) text = text.slice(0, -1);
  return text.length === 0 ? "." : text;
}

/** Absolute, normalised, `/`-separated — the form every path is compared in. */
export function absolute(path: string, cwd: string): string {
  return slashes(resolve(cwd, path));
}

export function relativeTo(rootFull: string, fileFull: string): string {
  const rel = slashes(relative(rootFull, fileFull));
  if (rel.startsWith("..") || isAbsolute(rel)) {
    throw new Error(`${fileFull} is not under ${rootFull}`);
  }
  return rel;
}

/** Relative even when outside: `../shared/x.ts` names a file the roots do not cover. */
export function relativeOrOutside(rootFull: string, fileFull: string): string {
  return slashes(relative(rootFull, fileFull));
}

/** Deepest common ancestor of absolute directories — one root, no absolute anchor. */
export function commonRoot(fullDirectories: readonly string[]): string {
  if (fullDirectories.length === 0) throw new Error("at least one source root is required");
  let common = fullDirectories[0] as string;
  for (const candidate of fullDirectories.slice(1)) {
    while (!isUnder(candidate, common)) {
      const parent = dirname(common);
      if (parent === common) return common;
      common = parent;
    }
  }
  return common;
}

export function isUnder(full: string, root: string): boolean {
  const rootWithSeparator = root.endsWith("/") || root.endsWith(sep) ? root : `${root}/`;
  return full === root || full.startsWith(rootWithSeparator);
}

File: extractors/elixir/lib/codegraph_elixir/paths.ex
defmodule CodegraphElixir.Paths do
  @moduledoc """
  Paths as the model writes them: `/`-separated on every OS, root-relative,
  compared by UTF-16 code unit. The file table is sorted by these strings, and
  a Windows `\\` or a case-folding comparison would change every surrogate.
  """

  def slashes(path), do: String.replace(path, "\\", "/")

  @doc "The typed path with `/` separators and no trailing slash — the header's `root`."
  def display(typed) do
    text = typed |> slashes() |> String.trim_trailing("/")
    if text == "", do: ".", else: text
  end

  @doc "Absolute, normalised, `/`-separated — the form every path is compared in."
  def absolute(path, cwd), do: path |> Path.expand(cwd) |> slashes()

  def relative_to(root_full, file_full) do
    rel = file_full |> Path.relative_to(root_full) |> slashes()

    if String.starts_with?(rel, "..") or Path.type(rel) == :absolute do
      raise ArgumentError, "#{file_full} is not under #{root_full}"
    end

    rel
  end

  @doc "Deepest common ancestor of absolute directories — one root, no absolute anchor."
  def common_root([first | rest]) do
    Enum.reduce(rest, first, fn candidate, common -> climb(candidate, common) end)
  end

  defp climb(candidate, common) do
    if under?(candidate, common) do
      common
    else
      parent = Path.dirname(common)
      if parent == common, do: common, else: climb(candidate, parent)
    end
  end

  def under?(full, root),
    do: full == root or String.starts_with?(full, String.trim_trailing(root, "/") <> "/")
end
